syntax = "proto3";
package user_identity;

import "google/protobuf/timestamp.proto";

service UserIdentityService {

  rpc CreateUser(CreateUserIdentityRequest)
      returns (CreateUserIdentityResponse) {}
  rpc GetUserByUUID(GetUserIdentityByUUIDRequest)
      returns (GetUserIdentityByUUIDResponse) {}
  rpc UpdateUser(UpdateUserIdentityRequest)
      returns (UpdateUserIdentityResponse) {}
  rpc DeleteUser (DeleteUserIdentityRequest) 
      returns (DeleteUserIdentityRequest){}

  rpc CreateBranchOffice(CreateBranchOfficeIdentityRequest)
      returns (CreateBranchOfficeIdentityResponse) {}
  rpc GetBranchOfficesByEmail(GetBranchOfficesByEmailRequest)
      returns (stream GetBranchOfficesByEmailResponse) {}
  rpc GetBranchOfficeByUUID(GetBranchOfficeIdentityByUUIDRequest)
      returns (GetBranchOfficeIdentityByUUIDResponse) {}
  rpc UpdateBranchOffice(UpdateBranchOfficeIdentityRequest)
      returns (UpdateBranchOfficeIdentityResponse) {}
  rpc DeleteBranchOffice (DeleteBranchOfficeIdentityRequest) 
      returns (DeleteBranchOfficeIdentityResponse){}

  rpc CreateOrganization(CreateOrganizationIdentityRequest)
      returns (CreateOrganizationIdentityResponse) {}
  rpc GetOrganizations(GetOrganizationsRequest)
      returns (stream GetOrganizationsResponse) {}
  rpc GetOrganizationByUUID(GetOrganizationIdentityByUUIDRequest)
      returns (GetOrganizationIdentityByUUIDResponse) {}
  rpc UpdateOrganization(UpdateOrganizationIdentityRequest)
      returns (UpdateOrganizationIdentityResponse) {}
  rpc DeleteOrganization (DeleteOrganizationIdentityRequest) 
      returns (DeleteOrganizationIdentityRequest){}

  rpc GetBranchOfficeAssociateList(GetBranchOfficeAssociateListRequest)
      returns (stream GetBranchOfficeAssociateListResponse) {}

  rpc GetKratosUserIdentity(GetKratosUserIdentityRequest)
      returns (GetKratosUserIdentityResponse) {}
}


message CreateUserIdentityRequest {
  string uuid = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  string role = 5;
  string branch_office_uuid = 6;
}

message CreateUserIdentityResponse {
  string uuid = 1;
}

message GetUserIdentityByUUIDRequest {
  string uuid = 1;
  bool visible = 2;
}

message GetUserIdentityByUUIDResponse { 
  string uuid = 1;
  string first_name = 2;
  string last_name = 3;
  string role = 4;
  string email = 5;

  google.protobuf.Timestamp created_at = 6;
  google.protobuf.Timestamp updated_at = 7;
  
  bool visible = 8;

  BranchOfficeMessage branch_office = 9;
}

message UpdateUserIdentityRequest {
  string uuid = 1;
  string first_name = 2;
  string last_name = 3;
  string role = 4;
  string email = 5;
  string branch_office_uuid = 6;
}

message UpdateUserIdentityResponse {
  string uuid = 1;
}

message DeleteUserIdentityRequest {
  string uuid = 1;
}

message DeleteUserIdentityResponse {
  string uuid = 1;
}


message CreateBranchOfficeIdentityRequest {
  string name = 1;
  string street = 2;
  string number = 3;
  string city = 4;
  string state = 5;
  string zip_code = 6;
  string organization_uuid = 7;
}

message CreateBranchOfficeIdentityResponse {
  string uuid = 1;
}

message GetBranchOfficeIdentityByUUIDRequest {
  string uuid = 1;
  bool visible = 2;
}

message GetBranchOfficeIdentityByUUIDResponse { 
  string uuid = 1;
  string name = 2;
  string street = 3;
  string number = 4;
  string city = 5;
  string state = 6;
  int32 zip_code = 7;
  string organization_uuid = 8;

  google.protobuf.Timestamp created_at = 9;
  google.protobuf.Timestamp updated_at = 10;
  
  bool visible = 11;

  repeated UserMessage users = 12;
  OrganizationMessage organization = 13;
}

message UpdateBranchOfficeIdentityRequest {
  string uuid = 1;
  string name = 2;
  string street = 3;
  string number = 4;
  string city = 5;
  string state = 6;
  string zip_code = 7;
}

message UpdateBranchOfficeIdentityResponse {
  string uuid = 1;
}

message DeleteBranchOfficeIdentityRequest {
  string uuid = 1;
}

message DeleteBranchOfficeIdentityResponse {
  string uuid = 1;
}


message CreateOrganizationIdentityRequest {
  string organization_domain = 1;
  string name = 2;
}

message CreateOrganizationIdentityResponse {
  string uuid = 1;
}

message GetOrganizationIdentityByUUIDRequest {
  string uuid = 1;
  bool visible = 2;
}

message GetOrganizationsRequest {}

message GetOrganizationsResponse {
  string uuid = 1;
  string organization_domain = 2;
  string name = 3;

  google.protobuf.Timestamp created_at = 4;
  google.protobuf.Timestamp updated_at = 5;
  
  bool visible = 6;

  repeated BranchOfficeMessage branch_offices = 7;
}

message GetOrganizationIdentityByUUIDResponse { 
  string uuid = 1;
  string organization_domain = 2;
  string name = 3;

  google.protobuf.Timestamp created_at = 4;
  google.protobuf.Timestamp updated_at = 5;
  
  bool visible = 6;

  repeated BranchOfficeMessage branch_offices = 7;
}

message UpdateOrganizationIdentityRequest {
  string uuid = 1;
  string organization_domain = 2;
  string name = 3;
}

message UpdateOrganizationIdentityResponse {
  string uuid = 1;
}

message DeleteOrganizationIdentityRequest {
  string uuid = 1;
}

message DeleteOrganizationIdentityResponse {
  string uuid = 1;
}


message GetBranchOfficeAssociateIdentityListRequest {
  string branch_office_uuid = 1;
  bool visible = 2;
}

message GetBranchOfficeAssociateIdentityListResponse {
  repeated UserMessage user_identity = 1;
}


message UserMessage {

  string uuid = 1;
  string first_name = 2;
  string last_name = 3;
  string role = 4;
  string email = 5;

  google.protobuf.Timestamp created_at = 6;
  google.protobuf.Timestamp updated_at = 7;
  
  bool visible = 8;
  
}

message BranchOfficeMessage {

  string uuid = 1;
  string name = 2;
  string street = 3;
  string number = 4;
  string city = 5;
  string state = 6;
  int32 zip_code = 7;

  google.protobuf.Timestamp created_at = 8;
  google.protobuf.Timestamp updated_at = 9;

  bool visible = 10;

  repeated UserMessage users = 11;

}

message OrganizationMessage {
  string uuid = 1;
  string organization_domain = 2;
  string name = 3;
}


message GetBranchOfficesByEmailRequest {
  string organization_domain = 1;
}

message GetBranchOfficesByEmailResponse {
  string uuid = 1;
  string name = 2;
}


message GetBranchOfficeAssociateListRequest {
  string uuid = 1;
}

message GetBranchOfficeAssociateListResponse {
  string email = 1;
}


message GetKratosUserIdentityRequest {
  string ory_kratos_cookie = 1;
  string type = 2;
}

message GetKratosUserIdentityResponse {
  string user_identity = 1;
}